heron/common/src/java/com/twitter/heron/common/basics/WakeableLooper.java

```diff
@@ -92,8 +92,16 @@ private void runOnce() {
   }
 
   private void onExit() {
-    for (Runnable r : exitTasks) {
-      r.run();
+    // Using CopyOnWriteArrayList to prevent from tasks in exitTasks from clearing
+    // the exitTasks list and potentially causing IndexOutOfBoundsException.
+    // The method iterator() in CopyOnWriteArrayList return an iterator object that holds
+    // different copy of the elements, hence the term snapshot iterator. The snapshot iterator
+    // doesn’t allow modifying the list while traversing, and it will not throw
+    // ConcurrentModificationException if the list is being modified by other thread during
+    // the traversal, and the read and write operations work on different copies of elements.
```

Comment:
two questions:
1. is modifying the list at the time expected? If not, maybe a more proper fix is to solve issue in the appending code?
2. will some exit tasks and tasks on wake up being missed with copyOnWrite list?

Reply:
IIRC, previously we could only add a task to such kind of `tasks-list` when executing tasks from it, but we could not remove a task.

On the other hand, WakeableLooper is designed not to be thread-safe; so it should be executed with only one thread.